core/asm/asm.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2014, 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// Provides support for dealing with EVM assembly instructions (e.g., disassembling them).
+package asm
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/core/vm"
+)
+
+// Apply function to each disassembled EVM instruction.
+func ForEachDisassembledInstruction(script []byte, fun func(uint64, vm.OpCode, []byte)) error {
```

Comment:
IMHO it would be better to have a real iterator here. The API would be as follows:

```go
it := asm.NewIterator(code)
for it.Next() { // Next returns true until there are no instructions left.
    // The PC, Op, Arg fields contain values for the current instruction.
    fmt.Println(it.PC, it.Op, it.Arg)
}
fmt.Println(it.Error()) // Error returns any error that occurred in Next.
```